L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+#include <memory>
+#include <string>
+
+static constexpr float mu_mass = 0.105658369;
+
+using namespace l1t;
+
+class L1TkMuonProducer : public edm::EDProducer {
+public:
+  typedef TTTrack<Ref_Phase2TrackerDigi_> L1TTTrackType;
+  typedef std::vector<L1TTTrackType> L1TTTrackCollectionType;
+
+  struct PropState {  //something simple, imagine it's hardware emulation
+    PropState() : pt(-99), eta(-99), phi(-99), sigmaPt(-99), sigmaEta(-99), sigmaPhi(-99), valid(false) {}
+    float pt;
+    float eta;
+    float phi;
+    float sigmaPt;
+    float sigmaEta;
+    float sigmaPhi;
+    bool valid;
+  };
+
+  enum AlgoType { kTP = 1, kDynamicWindows = 2, kMantra = 3 };
+
+  explicit L1TkMuonProducer(const edm::ParameterSet&);
+  ~L1TkMuonProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  PropState propagateToGMT(const L1TTTrackType& l1tk) const;
+  double sigmaEtaTP(const RegionalMuonCand& mu) const;
+  double sigmaPhiTP(const RegionalMuonCand& mu) const;
+
+  // the TP algorithm
+  void runOnMTFCollection_v1(const edm::Handle<RegionalMuonCandBxCollection>&,
+                             const edm::Handle<L1TTTrackCollectionType>&,
+                             TkMuonCollection& tkMuons,
+                             const int detector) const;
+
+  // algo for endcap regions using dynamic windows for making the match
+  void runOnMTFCollection_v2(const edm::Handle<EMTFTrackCollection>&,
+                             const edm::Handle<L1TTTrackCollectionType>&,
+                             TkMuonCollection& tkMuons) const;
+
+  // given the matching indexes, build the output collection of track muons
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               const edm::Handle<RegionalMuonCandBxCollection>& muonH,
+                               int detector) const;
+
+  // as above, but for the TkMu that were built from a EMTFCollection - do not produce a valid muon ref
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               int detector) const;
+
+  // dump and convert tracks to the format needed for the MAnTra correlator
+  std::vector<L1TkMuMantraDF::track_df> product_to_trkvec(const L1TTTrackCollectionType& l1tks);  // tracks
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const RegionalMuonCandBxCollection& l1mtfs);  // regional muon finder
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const EMTFTrackCollection& l1mus);  // endcap muon finder - eventually to be moved to regional candidate
+
+  float ETAMIN_;
```

Comment:
member variables should start with lower case

Reply:
done